{% extends "base.html" %}

<!-- {% block title %}Preference{% endblock %} -->

{% block content %}
<!-- <script src="scripts/scripts.js"></script> -->
    <!-- Page Content -->
    <div class="title-container">
        <div class="inner-title">
            <h1 class="page-title">Help</h1>
        </div>
    </div>

    <div class="container-help">
       <div class="help-box">

        <h2>Getting Started</h2>
            <p>Welcome to the Help page! Here, you'll find guidance on how to use the features of the app.</p>

            <p>You can Navigate through the various applications using the navigation bar at the top of the screen.
                This application contains three functions to analyze plasmid libraries.  </p>

            <div class="help-section">
                <h2>Variable Sites</h2>
                <p>The <strong>Find Variable Sites</strong> tool requires eight input parameters. Below is a detailed explanation of each input:</p>
    
                <p>
                    This tool processes a FASTQ library and generates an Excel spreadsheet with the counts of all combinations of variable sites detected. Sequences can also be further subdivided using the barcode option, which creates an additional tab in the spreadsheet that counts only the sequences containing the specified barcode.</p>
                <ul>
                    <li><strong>Reference Sequence (.fasta)</strong>
                        <p>This is a fasta file either in the (.fa) or (.fasta) format. The fasta file needs to be the reference sequence for the plasmid library.  
                            The reference sequence should include variable sites, which are marked using a non-standard codon, such as NNK or a similar format.  
                            The rest of the Reference Sequence should match the original plasmid sequence the library is based off of.  
                            This tool uses the non-standard nucleotide to locate the variable sites to be sampled (-,N,K,X. ect). 
                            Once identified it stores the nucleotides on the 5' and 3' end of the variable to identify the variables sites in the fastq library. 
                        </p>
                    </li>
                    <li><strong>Data Sequence (.fastq)</strong>
                        <p>This is the fastq file containing sequencing data from a plasmid library. 
                            The library should be derived from the reference sequence (provided above).  
                            The tool will use this file to compare against the reference to identify variable sites. 
                            The program cycles through the library twice for each variable site first identifying all forward strands to the reference sequence then taking the reverse complement of the reference sequence to match to the library. 
                            Be careful that the guide sequences used in the reference sequence reverse complement are not present anywhere on the forward strand. If this is unavoidable you can expand the range of sites, the program uses to identify variable sites or add a barcode sequence (see below). 
                        </p>
                    </li>
                    <li><strong>Output File Name</strong>
                        <p>Enter the desired name for the output file that will be generated after processing. This file will contain the results of the variable site identification. The file will be created in the Excel format (.xlsx).</p>
                    </li>
                    <li><strong>Number of Matching Nucleotides (5' End)</strong>
                        <p>This value specifies how many nucleotides, starting from the 5' end of variable the sequence, need to match the reference sequence in order for a site to be considered a variable site. Any site that does not match this exact region will be discarded.</p>
                    </li>
                    <li><strong>Number of Matching Nucleotides (3' End)</strong>
                        <p>Similar to the 5' end, this value determines how many nucleotides, starting from the 3' end, must match the reference sequence to confirm the site as a variable site. Sites that do not match this region will be excluded.</p>
                    </li>
                    <li><strong>Phred Score Threshold</strong>
                        <p>This threshold determines the minimum Phred score required for each nucleotide within the matching 5' and 3' regions. If any nucleotide has a Phred score below this threshold, the corresponding site will be discarded as unreliable. If the optional barcode option is used the barcode will use this Phred score value. </p>
                    </li>
                    <li><strong>Number of Variable Sites</strong>
                        <p>This field specifies the expected number of variable sites in the reference sequence. If the number of variable sites detected does not match this expectation, the program will return an error, indicating a mismatch.</p>
                    </li>
                    <li><strong>(Optional) Barcode Check</strong>
                        <p>If you provide a nucleotide barcode in this field, the program will filter out any sequences that do not contain the specified barcode. This is useful for ensuring that only sequences with the correct barcode are considered in the analysis. If multiple Bar codes are placed below in a comma separated list the program will create a separate Excel tab in the output file with each tab only accounting for sequences that match the barcode.
                            
                            Example (ATTGGCC,TTGGCCCC,TTTTTAAG). </p>
                    </li>
                </ul>
            
            
            

            <div class="help-section">
				<h2>Amino Acid Preference</h2>
				<p>The <strong>Find Amino Acid Preference</strong> tool requires eight input parameters. Below is a detailed explanation of each:</p>

                <p>This tool compares the change in amino acid preferences between a pre-mutation and a post-mutation library at a specific site. It normalizes the mutations against the wild-type sequences for accurate analysis. The output format is an excel file with the normalized amino acid prefrences.</p>
				<ul>
					<li><strong>Reference Sequence (.fasta)</strong>
						<p>This is a fasta file either in the (.fa) or (.fasta) format. The fasta file needs to be the reference sequence for the plasmid library. The reference sequence should include variable sites, which are marked using a non-standard codon, such as NNK or a similar format. The rest of the Reference Sequence should match the original plasmid sequence the library is based off of. This tool uses the non-standard nucleotide to locate the variable sites to be sampled (-,N,K,X. ect). Once identified it stores the nucleotides on the 5' and 3' end of the variable to identify the variables sites in the fastq library.</p>
					</li>
					<li><strong>Data Sequence 1(.fastq)</strong>
						<p>This is the fastq file containing sequencing data from the pre mutant plasmid library. The library should be derived from the reference sequence (provided above). The tool will use this file to compare against the reference to identify variable sites. The program cycles through the library twice for each variable site first identifying all forward strands to the reference sequence then taking the reverse complement of the reference sequence to match to the library. Be careful that the guide sequences used in the reference sequence reverse complement are not present anywhere on the forward strand. If this is unavoidable you can expand the range of sites, the program uses to identify variable sites or add a barcode sequence (see below).</p>
					</li>
                    <li><strong>Data Sequence 2(.fastq)</strong>
                        <p>This is the fastq file requires the same format as Data Sequence 1. Data Sequence 2 should be sequenced from the result of mutating the plasmids generated Data Sequence 1.</p>
                    </li>
					<li><strong>Number of Matching Nucleotides (5' End)</strong>
						<p>This value specifies how many nucleotides, starting from the 5' end of variable the sequence, need to match the reference sequence in order for a site to be considered a variable site. Any site that does not match this exact region will be discarded.</p>
					</li>
					<li><strong>Number of Matching Nucleotides (3' End)</strong>
						<p>Similar to the 5' end, this value determines how many nucleotides, starting from the 3' end, must match the reference sequence to confirm the site as a variable site. Sites that do not match this region will be excluded.</p>
					</li>
					<li><strong>Phred Score Threshold</strong>
						<p>This threshold determines the minimum Phred score required for each nucleotide within the matching 5' and 3' regions. If any nucleotide has a Phred score below this threshold, the corresponding site will be discarded as unreliable. If the optional barcode option is used the barcode will use this Phred score value.</p>
					</li>
					<li><strong>Output File Name</strong>
						<p>Enter the desired name for the output file that will be generated after processing. This file will contain the results of the variable site identification. The file will be created in the Excel format (.xlsx).</p>
					</li>
                    <li><strong>Wild type amino acid</strong>
                        <p>Enter the wild type amino acid present at the variable site.</p>
					</li>
				</ul>
            
            </div>
        </div>

        <div class="help-section">
            <h2>Build primer</h2>
            <p>The <strong>Build primer</strong> tool requires two input parameters. Below is a detailed explanation of each:</p>
            <p>The tool creates the 4 primers needed to create a plasmid library from a reference sequence. 

                The tool creates the 4 primers needed to create a plasmid library from a reference sequence in a text file (txt). 
                Forward Primer 1 (F1): Primer at the very beginning of the sequence needed for PCR on the full sequence. 
                Forward Primer 2 (F2): Primer after the 3' end of the variable site without overlapping the site. 
                Forward Primer 1 (F1): The 60 base pair primer that covers the variable sequence. 
                Reverse Primer 2 (R2): Primer at the end of sequence needed for PCR on the full sequence. 
                 
                Any sequences with basic self-binding or other errors will be flagged at the end of the text file. 

            </p>
            <ul>
            <li><strong>Reference Sequence (.fasta)</strong>
                <p>
                This is the fastq file containing sequencing data from a plasmid library. 
                The library should be derived from the reference sequence (provided above).  
                The tool will use this file to compare against the reference to identify variable sites. 
                </p>
            </li>
            <li><strong>Number of Variable Sites</strong>
                <p>This field specifies the expected number of variable sites in the reference sequence. If the number of variable sites detected does not match this expectation, the program will return an error, indicating a mismatch.</p>
            </li>
            </ul>
        </div>


    </div>

    <script>
        document.addEventListener("DOMContentLoaded", function() {
            var helpSections = document.querySelectorAll('.help-box .help-section h2');
            helpSections.forEach(function(header) {
                header.addEventListener('click', function() {
                    this.parentElement.classList.toggle('active');
                    let content = this.nextElementSibling;
                    if (content.style.display === "block") {
                        content.style.display = "none";
                    } else {
                        content.style.display = "block";
                    }
                });
            });
        });
    </script>

{% endblock %}
